src/shared/game/servergamestate.cpp

```diff
@@ -0,0 +1,233 @@
+#include "shared/game/servergamestate.hpp"
+#include "shared/game/sharedgamestate.hpp"
+
+/*	Constructors and Destructors	*/
+
+ServerGameState::ServerGameState(GamePhase start_phase, GameConfig config) {
+	this->phase = start_phase;
+	this->timestep = FIRST_TIMESTEP;
+	this->timestep_length = config.game.timestep_length_ms;
+	this->lobby.max_players = config.server.max_players;
+
+	//	Initialize SmartVectors with max sizes
+	this->objects = SmartVector<Object*>(MAX_NUM_OBJECTS);
+
+	this->base_objects = SmartVector<Object>(MAX_NUM_BASE_OBJECTS);
+}
+
+ServerGameState::ServerGameState(GamePhase start_phase) {
+	this->phase = start_phase;
+	this->timestep = FIRST_TIMESTEP;
+	this->timestep_length = TIMESTEP_LEN;
+	this->lobby.max_players = MAX_PLAYERS;
+
+	//	Initialize SmartVectors with max sizes
+	this->objects = SmartVector<Object*>(MAX_NUM_OBJECTS);
+
+	this->base_objects = SmartVector<Object>(MAX_NUM_BASE_OBJECTS);
+}
+
+ServerGameState::ServerGameState() : ServerGameState(GamePhase::LOBBY) {}
+
+ServerGameState::~ServerGameState() {}
+
+/*	SharedGameState generation	*/
+SharedGameState ServerGameState::generateSharedGameState() {
+	//	Create a new SharedGameState instance and populate it with this
+	//	ServerGameState's data
+	SharedGameState shared;
+
+	//	Initialize object vector
+	for (int i = 0; i < this->objects.size(); i++) {
+		Object** ptrToPtr = this->objects.get(i);
+
+		if (ptrToPtr == nullptr) {
+			//	Push empty object to SharedObject SmartVector
+			shared.objects.pushEmpty();
+		}
+		else {
+			Object* object = *ptrToPtr;
+			
+			//	Create a SharedObject from this object
+			shared.objects.push(object->generateSharedObject());
+		}
```

Comment:
Handling the double pointer to an object here feels kinda clunky.

Maybe we can have `SmartVector::get` return a `std::shared_ptr` to the underlying Object (this would mean that we'd have a `SmartVector` of `Object` not `Object*`).

Also, this isn't that big of a deal and it's totally fine if we keep this.